doc/connectivity/networking/api/socket_service.rst

```diff
@@ -0,0 +1,142 @@
+.. _socket_service_interface:
+
+Socket Service
+##############
+
+.. contents::
+    :local:
+    :depth: 2
+
+Overview
+********
+
+The socket service API can be used to install a handler that is called if there
+is data received to a socket. The API helps to avoid creating a dedicated thread
+for each TCP or UDP services that the application provides. Instead one thread
+is created which can serve data to multiple listening sockets which saves memory
+as only one thread needs to be created in the system in this case.
+
+See :zephyr:code-sample:`sockets-service-echo` sample application to learn how
+to create a simple BSD socket based server application using the sockets service API.
+The source code for this sample application can be found at:
+:zephyr_file:`samples/net/sockets/echo_service`.
+
+API Description
+***************
+
+Socket service API is enabled using :kconfig:option:`CONFIG_NET_SOCKETS_SERVICE`
+config option and implements the following operations:
+
+* :c:macro:`NET_SOCKET_SERVICE_SYNC_DEFINE`
+
+  Define a network socket service. This socket service is created with extern
+  scope so it can be used from multiple C source files.
+
+* :c:macro:`NET_SOCKET_SERVICE_SYNC_DEFINE_STATIC`
+
+  Define a network socket service with static scope. This socket service can only
+  be used within one C source file.
+
+* :c:func:`net_socket_service_register`
+
+  Register pollable sockets for this service. User must create the sockets
+  before this call.
+
+* :c:func:`net_socket_service_unregister`
+
+  Remove pollable sockets for this service. User can close the sockets after
+  this call.
+
+* :c:type:`net_socket_service_handler_t`
+
+  User specified callback which is called when data is received to the
+  listening socket.
+
+Application Overview
```

Comment:
The overview really only shows a "simple" UDP service, while it is more complex for TCP as it creates new socket file descriptors for connecting clients.
It would be nice to have that distinction in the documentation.

Reply:
Yes, I thought about this, but did not want to write complete application as the `echo-service` application has that implemented. But we could put something more about TCP and what to do with the accepted socket.

Reply:
Yes, one thing that tricked me, and would be nice to have documented, is that the `accept` call can be done in the callback too. This isn't the case in the `echo-service` sample, where instead a loop is used for a single client.

In most cases, I think, services should serve multiple clients.

Reply:
Sure, we could enhance the echo-service sample to be a bit more realistic. Patches are welcome to enhance the sample.